### DISCUSSION TITLE
Refactoring settings/config/environment variable management -- when and how?

### DISCUSSION THREAD

**Discussion by cjohnhanson** [meltano/meltano#7809] [General]
Settings management and environment variable expansion has been a consistent pain point and source of bugs within meltano core. We've made some progress--particularly with regards to better testing and all the bugfixes done as part of the meltano manifest work--over the past year, but it's often one step forward and two steps back as bugfixes and changes inevitably cause new bugs. I've increasingly been leaning towards doing a full refactor, possibly leaning on a third party library/framework like [hyedra](https://hydra.cc/docs/intro/) or [Dynaconf](https://www.dynaconf.com/). It's not feasibly at this point to do such a significant refactor as part of our 3.0 launch, but taking this on as part of a major version bump would be definitely be preferable since there's almost definitely going to be changes which break existing users' `meltano.yml` or otherwise changes the existing behavior. But on the other hand, this is a pressing issue for our users and likely not one we want to wait until a 4.0 release to handle.

I'm interested in gathering everyone's thoughts about how best to approach this and what a potential timeline for a significant refactor could look like, whether that's as part of a 4.0 release or if there's an approach we can take that would be guaranteed to only make changes that could be released in a minor version bump. 

Related slack thread: https://meltano.slack.com/archives/C03GKHWS0HM/p1685131341359449

CC: @meltano/engineering

**Comment by WillDaSilva** [comment]
@cjohnhanson You and I have talked a fair bit about this topic, so you're already aware that I'm keen for us to tackle a refactor of how Meltano handles configuration for myriad reasons. The most concrete suggestion I've made to date on this topic is:
- https://github.com/meltano/meltano/issues/7400

In a nutshell, the above suggests using the library `datafiles` to handle Meltano config files in a way that would be, ideally, fully backwards compatible.

The big (related) questions I have on my mind when considering these approaches towards refactoring Meltano's config management are:
- What are our primary objections? These could include simplifying the code, fixing known bugs, making it easier to add new features, or straight up adding new features. 
- How backwards compatible will it be?
    - Will the resulting config management system warrant a major version bump?
    - Should we finally bump the `version` attribute in `meltano.yml` to something other than `1`?
        - If we do so, should we continue to support the old config format in some way, or simply print an error along the lines of "you must use a version of Meltano older than 4.0 to use the `meltano.yml` file, or upgrade your Meltano program to the new config format"?
- How much of an impact on Meltano (the CLI application) do we want this change to have?
    - Would we, for instance, use Dynaconf's built-in secrets management? Or it's `@format` feature? Etc. These would be product decisions, for the most part.
    - Could we begin to automatically generate the JSONSchema, thereby forevermore preventing its drift away from what Meltano actually does?
- How much work will it be?
    - How amenable are the libraries in question to doing the particular things we want them to do? For instance, the `datafiles` library doesn't support using multiple files with `include_paths` as we currently do, but it should be possible to extend the library with support for that.

In considering the scope of the refactor we have to perform a balancing act. On one hand the refactor must be large/significant enough that it can actually root out the underlying problems with the existing approaches and address them, rather than merely add on another layer of complexity. On the other hand we need to avoid making this such a large and daunting piece of work that it's impractical to ever accomplish. Countless refactors/rewrites have failed completely because they could not be completed in an iterative fashion, so while keeping in mind that the refactor ought to produce a significant and foundational change, we should aim to design and implement a minimum viable refactor that can be built upon sustainably.

With all that said, as tempting as starting from scratch with the config format/design would be (with a bump to the `version` attribute of `meltano.yml`), I think we should aim for backwards compatibility, and make this primarily a internal change. Our users will notice the fixed bugs, and the performance improvements (or rather, they'll _not_ notice bugs and poor performance).

In summary:
- I agree that a refactor in this area should be done.
- My main objectives for (and what is driving my prioritization of) such a refactor are: addressing existing bugs, and preventing new bugs.
- I'd like to make the refactor preserve backwards compatibility, and not require a major version bump.
- I don't think Dynaconf or Hyedra will be well-suited for this refactor because, as far as I can tell, using them would necessitate a significantly larger refactor than what we could get away with without them, and using them would likely result in large user-facing changes that would necessitate a major version bump and migration strategy.
